### PR TITLE
Hide compiler generated branches for try/catch blocks inside async state machine

### LINKED ISSUE #465: Closing bracket uncovered in try-catch statement when exception is re-thrown in catch
The closing bracket is showing up as uncovered when I re-throw an exception from a `catch` statement. Is this expected behavior? The closing bracket is technically unreachable code, and it prevents me from reaching 100% coverage on the method when coverlet says it is uncovered by a test.

```
try
{
...
}
catch (Exception)
{
    // handle exception (retry/log/etc.)

    throw;
} // ** uncovered line **
```

### PR DESCRIPTION
**Pull request by matteoerigozzi**
Fixes #465

Improve compiler generated branches detection inside MoveNext() method of async state machine.

This core remove all unexpected branches
image

### CODE REVIEW

**Comment by matteoerigozzi** [comment]
Last update removes unexpected uncovered line after exception re-thrown inside catch block in the async state machine.
image

File: src/coverlet.core/Instrumentation/Instrumenter.cs [hunk 1]
```diff
@@ -447,6 +447,42 @@ private void InstrumentIL(MethodDefinition method)
                 var sequencePoint = method.DebugInformation.GetSequencePoint(instruction);
                 var targetedBranchPoints = branchPoints.Where(p => p.EndOffset == instruction.Offset);
 
+                /*
+                    Skip instrumentation after exception re-throw insiede catch block (only for async state machine MoveNext())
```
Comment (MarcoRossignoli) [inline src/coverlet.core/Instrumentation/Instrumenter.cs]:
```suggestion
                    Skip instrumentation after exception re-throw inside catch block (only for async state machine MoveNext())
```

File: src/coverlet.core/Instrumentation/Instrumenter.cs [hunk 2]
```diff
@@ -447,6 +447,42 @@ private void InstrumentIL(MethodDefinition method)
...
+		            // (no C# code)
+		            IL_00d2: nop
+		            IL_00d3: nop
+                    ...
+                */
+                Instruction previous;
+                if (
+                        CecilSymbolHelper.IsMoveNextInsideAsyncStateMachine(method) &&
+                        (previous = CecilSymbolHelper.GetPreviousNoNopInstruction(instruction)) != null &&
+                        previous != null &&
```
Comment (MarcoRossignoli) [inline src/coverlet.core/Instrumentation/Instrumenter.cs]: double check?Null check is on above line, right?

File: src/coverlet.core/Instrumentation/Instrumenter.cs [hunk 3]
```diff
@@ -447,6 +447,42 @@ private void InstrumentIL(MethodDefinition method)
...
+		            IL_00cd: callvirt instance void [System.Runtime]System.Runtime.ExceptionServices.ExceptionDispatchInfo::Throw()
+		            // (no C# code)
+		            IL_00d2: nop
+		            IL_00d3: nop
+                    ...
+                */
+                Instruction previous;
+                if (
+                        CecilSymbolHelper.IsMoveNextInsideAsyncStateMachine(method) &&
+                        (previous = CecilSymbolHelper.GetPreviousNoNopInstruction(instruction)) != null &&
```
Comment (MarcoRossignoli) [inline src/coverlet.core/Instrumentation/Instrumenter.cs]: If we're sure about a `nop` why don't check for it for better performace?Or are you not sure?

File: src/coverlet.core/Symbols/CecilSymbolHelper.cs [hunk 4]
```diff
@@ -202,6 +202,89 @@ Single branch
...
+
+                            IL_0049: nop                                <- start exception handler code
+                            ...
+
+                            
+                            So starting from branch instruction 'beq.s', we can go back to starting block instruction 
+                            which is always 5 step before and then check if this istruction is the end of an exception handler block
+                        */
+
+                        int branchIndex = instructions.IndexOf(instruction);
```
Comment (MarcoRossignoli) [inline src/coverlet.core/Symbols/CecilSymbolHelper.cs]: Can you use binary search?IL instruction offset are sequential ordered O(n) -> O(log n) similar https://github.com/tonerdo/coverlet/blob/master/src/coverlet.core/Symbols/CecilSymbolHelper.cs#L448

File: src/coverlet.core/Symbols/CecilSymbolHelper.cs [hunk 5]
```diff
@@ -202,6 +202,89 @@ Single branch
...
+                        if (
+                                branchIndex >= 5 &&
+                                instructions[branchIndex - 5].OpCode == OpCodes.Ldarg &&
```
Comment (MarcoRossignoli) [inline src/coverlet.core/Symbols/CecilSymbolHelper.cs]: I feel more safe if we check array bounds

File: src/coverlet.core/Symbols/CecilSymbolHelper.cs [hunk 6]
```diff
@@ -279,6 +362,13 @@ Single branch
             return true;
         }
 
+        internal static Instruction GetPreviousNoNopInstruction(Instruction current)
+        {
+            Instruction instruction = current.Previous;
+            for (; instruction != null && instruction.OpCode == OpCodes.Nop; instruction = instruction.Previous) { }
```
Comment (MarcoRossignoli) [inline src/coverlet.core/Symbols/CecilSymbolHelper.cs]: I don't like empty body can you invert loop condition and return from inside the body?

File: src/coverlet.core/Symbols/CecilSymbolHelper.cs [hunk 7]
```diff
@@ -202,6 +202,89 @@ Single branch
...
+                            IL_00bb: ldloc.s 6
+                            // (no C# code)
+                            IL_00bd: brtrue.s IL_00c6
+
+                            So we can go back to previous instructions and skip this branch if recognize that type of code block
+                        */
+
+                        if (
+                                branchIndex >= 3 && // avoid out of range exception (need almost 3 instruction before the branch)
+                                instructions[branchIndex - 3].OpCode == OpCodes.Isinst &&
```
Comment (MarcoRossignoli) [inline src/coverlet.core/Symbols/CecilSymbolHelper.cs]: I'd add bounds also here if for some reason we fail to get index we go on with no problem, or are you sure about every possible case?

File: test/coverlet.core.tests/Coverage/CoverageTests.CatchBlock.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+                        ((Task)instance.TestAsync()).ConfigureAwait(false).GetAwaiter().GetResult();
+
+                        return Task.CompletedTask;
+                    }, persistPrepareResultToFile: pathSerialize);
+                    return 0;
+                }, path).Dispose();
+
+                CoverageResult result = TestInstrumentationHelper.GetCoverageResult(path);
+
+                result.Document("Instrumentation.CatchBlock.cs")
```
Comment (MarcoRossignoli) [inline test/coverlet.core.tests/Coverage/CoverageTests.CatchBlock.cs]:
```suggestion
                TestInstrumentationHelper.GetCoverageResult(path).Document("Instrumentation.CatchBlock.cs")
```

File: test/coverlet.core.tests/Coverage/CoverageTests.CatchBlock.cs [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+                    {
+                        instance.Test();
+                        ((Task)instance.TestAsync()).ConfigureAwait(false).GetAwaiter().GetResult();
+
+                        return Task.CompletedTask;
+                    }, persistPrepareResultToFile: pathSerialize);
+                    return 0;
+                }, path).Dispose();
+
+                CoverageResult result = TestInstrumentationHelper.GetCoverageResult(path);
```
Comment (MarcoRossignoli) [inline test/coverlet.core.tests/Coverage/CoverageTests.CatchBlock.cs]:
```suggestion
                
```

File: test/coverlet.core.tests/Coverage/CoverageTests.CatchBlock.cs [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+                result.Document("Instrumentation.CatchBlock.cs")
+                    .AssertLinesCoveredAllBut(BuildConfiguration.Debug,
```
Comment (MarcoRossignoli) [inline test/coverlet.core.tests/Coverage/CoverageTests.CatchBlock.cs]:
```suggestion
                    .AssertLinesCoveredAllBut(BuildConfiguration.Debug, 41, 51)
```

File: test/coverlet.core.tests/Coverage/CoverageTests.CatchBlock.cs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+                        41,
```
Comment (MarcoRossignoli) [inline test/coverlet.core.tests/Coverage/CoverageTests.CatchBlock.cs]:
```suggestion

```

File: test/coverlet.core.tests/Coverage/CoverageTests.CatchBlock.cs [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+                    }, persistPrepareResultToFile: pathSerialize);
+                    return 0;
+                }, path).Dispose();
+
+                CoverageResult result = TestInstrumentationHelper.GetCoverageResult(path);
+
+                result.Document("Instrumentation.CatchBlock.cs")
+                    .AssertLinesCoveredAllBut(BuildConfiguration.Debug,
+                        41, 
+                        51)
```
Comment (MarcoRossignoli) [inline test/coverlet.core.tests/Coverage/CoverageTests.CatchBlock.cs]:
```suggestion
```

File: test/coverlet.core.tests/Samples/Instrumentation.CatchBlock.cs [hunk 13]
```diff
@@ -0,0 +1,55 @@
+﻿// Remember to use full name because adding new using directives change line numbers
+
+using System.Threading.Tasks;
+
+namespace Coverlet.Core.Samples.Tests
+{
+    public class CatchBlock
```
Comment (MarcoRossignoli) [inline test/coverlet.core.tests/Samples/Instrumentation.CatchBlock.cs]: I'd add more sample
* Multiple throw inside catch with some condition i.e.
```cs
if (bool) 
    throw; 
else
   throw new Exception();
```
* Nested try/catch
* Catch with Type `catch(DivideByZero)` and inside condition with `throw` and `throw Exception` single and nested

**Comment by MarcoRossignoli** [comment]
The idea generally is ok to me, we need to be sure that if heuristics fail instrumentation go on, so some more check is not a great problem...I prefer worste perf than fail.

**Comment by MarcoRossignoli** [comment]
Please can you attach a full IL source of `MoveNext` to better review?I'd like to do another review round.

**Comment by matteoerigozzi** [comment]
@MarcoRossignoli PTAL

I improved the heuristic logic and added several tests to validate it.

**Comment by matteoerigozzi** [comment]
Attach generated IL for MoveNext() of this example (nested try/catch blocks with conditional exception rethrown)

image

[MoveNext.txt](https://github.com/tonerdo/coverlet/files/4230260/MoveNext.txt)

**Comment by MarcoRossignoli** [comment]
I'm checking...I need some time, this fix is "articulated", thank's

**Comment by MarcoRossignoli** [comment]
@matteoerigozzi can you rebase?

**Comment by matteoerigozzi** [comment]
@MarcoRossignoli rebased!

File: src/coverlet.core/Symbols/CecilSymbolHelper.cs [hunk 14]
```diff
@@ -431,6 +599,96 @@ public static List<BranchPoint> GetBranchPoints(MethodDefinition methodDefinitio
...
+            {
+                return true;
+            }
+
+            // find the caller of current instruction and detect if not coverable
+            prev = instruction.Previous;
+            while (prev != null)
+            {
+                if (prev.Operand is Instruction i && (i.Offset == instruction.Offset || i.Offset == prev.Next.Offset)) // caller
+                //if (prev.Operand is Instruction i && i.Offset == instruction.Offset) // caller
```
Comment (MarcoRossignoli) [inline src/coverlet.core/Symbols/CecilSymbolHelper.cs]: Can be removed?

File: src/coverlet.core/Instrumentation/Instrumenter.cs [hunk 15]
```diff
@@ -447,6 +447,14 @@ private void InstrumentIL(MethodDefinition method)
                 var sequencePoint = method.DebugInformation.GetSequencePoint(instruction);
                 var targetedBranchPoints = branchPoints.Where(p => p.EndOffset == instruction.Offset);
 
+                // Check if the instruction is coverable
+                if (CecilSymbolHelper.IsMoveNextInsideAsyncStateMachine(method) &&
+                    CecilSymbolHelper.IsNotCoverableInstructionAfterExceptionThrown(instruction))
```
Comment (MarcoRossignoli) [inline src/coverlet.core/Instrumentation/Instrumenter.cs]: Can you fold this two method in one and hide `IsMoveNextInsideAsyncStateMachine` that is used inside `CecilSymbolHelper`.
Something like `SkipInstructionAfterThrownInsideMoveNextAsyncStateMachine`

File: src/coverlet.core/Symbols/CecilSymbolHelper.cs [hunk 16]
```diff
@@ -252,6 +410,11 @@ public static List<BranchPoint> GetBranchPoints(MethodDefinition methodDefinitio
                         continue;
                     }
 
+                    if (compilerGeneratedBranchesToExclude.Contains(instruction.Offset))
```
Comment (MarcoRossignoli) [inline src/coverlet.core/Symbols/CecilSymbolHelper.cs]: Is it too complex avoid to generate the list above but check directly when we come to the branch instruction to exclude?

**Comment by MarcoRossignoli** [comment]
@matteoerigozzi have you tried to generate a report in "release"(some nop should disappear) and check if makes sense?

File: src/coverlet.core/Symbols/CecilSymbolHelper.cs (hunk 16 above)
Comment (matteoerigozzi) [reply src/coverlet.core/Symbols/CecilSymbolHelper.cs]: To generate this list we need to parse all method instructions, so I prefer to do that before the foreach (I thought that in the future we may have other branches to load in that list). If you prefer, I can move the code in a Skip style method and cache the list.. something like:
```csharp
SkipCompilerGeneratedBranchesForExceptionHandlersInsideAsyncStateMachine(Instruction instruction)
{

    if (CompilerGeneratedBranchesToExclude == null)
    {
         // Detect branches to exclude
    }

    return compilerGeneratedBranchesToExclude.Contains(instruction.Offset);
}
```

Comment (MarcoRossignoli) [reply src/coverlet.core/Symbols/CecilSymbolHelper.cs]: Yep better

Comment (matteoerigozzi) [reply src/coverlet.core/Symbols/CecilSymbolHelper.cs]: Not so better maybe. The list is for method, so is not useful to use static cache.  How can I refactor this code?

Comment (matteoerigozzi) [reply src/coverlet.core/Symbols/CecilSymbolHelper.cs]: @MarcoRossignoli please check last update

**Pull request merged**